File: crates/pitch-core-py/python/quickstart.py
"""Minimal pitch-core-py example: load a WAV, run SWIPE (pure DSP),
print voiced frames.

Run:

    # Build the Rust extension into the current Python env (requires `pip
    # install maturin` first):
    cd <repo-root>
    maturin develop -m crates/pitch-core-py/Cargo.toml --release

    # Then:
    pip install soundfile numpy
    python crates/pitch-core-py/python/quickstart.py path/to/voice.wav

For neural backends (CREPE, RMVPE, SwiftF0, FCPE), build with the
`onnx` feature instead:

    maturin develop -m crates/pitch-core-py/Cargo.toml --release \\
        --features onnx

then pass `algorithm="rmvpe"` (or any other) and `model_path=...` to
`PitchTracker(...)`. See `pitch_core_py.available_backends` for the
list compiled into your build.
"""

from __future__ import annotations

import sys
from pathlib import Path

import numpy as np
import soundfile as sf

import pitch_core_py as pcp


def main() -> int:
    if len(sys.argv) < 2:
        print("usage: quickstart.py <wav-file>", file=sys.stderr)
        return 1

    path = Path(sys.argv[1])
    audio, sr = sf.read(str(path), dtype="float32")
    if audio.ndim > 1:
        audio = audio.mean(axis=1).astype(np.float32)

    print(f"loaded {path}: {len(audio)/sr:.2f} s @ {sr} Hz", file=sys.stderr)
    print(f"available backends: {pcp.available_backends}", file=sys.stderr)

    # Pure-DSP SWIPE — no model file, no GPU. ~96.4% RPA on MIR-1K.
    tracker = pcp.PitchTracker(
        algorithm="swipe",
        input_sample_rate=sr,
        resample_chunk=1024,
        swipe_max_window=8192,
    )

    result = tracker.process(audio)
    pitch_hz = result["pitch_hz"]
    confidence = result["confidence"]
    timestamps = result["timestamps_s"]

    voiced = confidence >= 0.3
    n_voiced = int(voiced.sum())
    print(
        f"\n{len(pitch_hz)} frames, {n_voiced} voiced "
        f"({100*n_voiced/max(len(pitch_hz), 1):.1f}%)"
    )

    print("\nfirst 10 voiced frames:")
    voiced_idx = np.where(voiced)[0][:10]
    for i in voiced_idx:
        print(f"  {timestamps[i]:6.3f}s  {pitch_hz[i]:6.1f} Hz  "
              f"conf={confidence[i]:.2f}")

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/pitch-core/examples/quickstart.rs
//! Minimal pitch-core example: load a WAV, run SWIPE, print voiced frames.
//!
//! Pure-DSP — no model files, no `ort` / ONNX Runtime, ~2 MB binary.
//!
//! Run:
//!     cargo run --example quickstart --release -- path/to/voice.wav
//!
//! For neural backends (CREPE/RMVPE/SwiftF0/FCPE) see the
//! `pitch-core-onnx` crate's `bench` example.

use hound::WavReader;
use pitch_core::{PitchTracker, SwipeEstimator};
use std::env;
use std::error::Error;

fn main() -> Result<(), Box<dyn Error>> {
    let path = env::args()
        .nth(1)
        .ok_or("usage: quickstart <wav-file>")?;

    let mut reader = WavReader::open(&path)?;
    let spec = reader.spec();
    let sr = spec.sample_rate;
    let n_ch = spec.channels as usize;

    // Decode to f32 mono.
    let raw: Vec<f32> = match spec.sample_format {
        hound::SampleFormat::Float => reader
            .samples::<f32>()
            .collect::<Result<_, _>>()?,
        hound::SampleFormat::Int => {
            let max = ((1i64 << (spec.bits_per_sample - 1)) - 1) as f32;
            reader
                .samples::<i32>()
                .map(|s| s.map(|x| x as f32 / max))
                .collect::<Result<_, _>>()?
        }
    };
    let mono: Vec<f32> = if n_ch == 1 {
        raw
    } else {
        raw.chunks_exact(n_ch)
            .map(|c| c.iter().sum::<f32>() / n_ch as f32)
            .collect()
    };

    println!(
        "loaded {}: {:.2}s @ {} Hz",
        path,
        mono.len() as f32 / sr as f32,
        sr,
    );

    // Default SWIPE = Balanced preset (8192 max_window, ~170 ms latency,
    // 96.4% RPA on MIR-1K). For lower latency: SwipeEstimator::with_max_window(4096).
    let est = SwipeEstimator::new()?;
    let mut tracker = PitchTracker::new(est, sr, 1024)?;

    let frames = tracker.process(&mono)?;
    let voiced: Vec<_> = frames.iter().filter(|f| f.confidence >= 0.3).collect();

    println!(
        "{} frames, {} voiced ({:.1}%)",
        frames.len(),
        voiced.len(),
        100.0 * voiced.len() as f32 / frames.len().max(1) as f32,
    );

    println!("\nfirst 10 voiced frames:");
    for f in voiced.iter().take(10) {
        println!(
            "  {:6.3}s  {:6.1} Hz  conf={:.2}",
            f.time_s, f.pitch_hz, f.confidence,
        );
    }

    Ok(())
}
